### ISSUE TITLE
ERC-1726: Dividend-Paying Token Standard

### ISSUE THREAD

**Issue by Roger-Wu** [ethereum/EIPs#1726]
```
eip: <to be assigned>
title: Dividend-Paying Token Standard
author: Roger Wu (@Roger-Wu) <gsetcrw@gmail.com>, Tom Lam (@erinata) <tomlam@uchicago.edu>
type: Standards Track
category: ERC
status: Draft
created: 2019-01-27
```

## Simple Summary

A standard interface for dividend-paying tokens.

## Abstract

The following standard allows for the implementation of a standard API for dividend-paying tokens within smart contracts.
This standard provides basic functionality to allow anyone to pay and distribute ether to all token holders.
A token holder can then view or withdraw the ether distributed to them from the contract.

## Motivation

While the number of tokens and dapps with dividend paying features is growing, there is few widely adopted standard interfaces for such contracts.
A standard interface allows wallet/exchange applications to work with any dividend-paying token on Ethereum, which makes it much easier to view and manage dividends from different sources.

In this EIP, we propose a standard interface and a well-tested implementation for a dividend-paying token contract.
We hope that our work could help the community to reach a consensus of the standard.

## Specification

```solidity
pragma solidity ^0.5.0;


/// @title Dividend-Paying Token Interface
/// @author Roger Wu (https://github.com/roger-wu)
/// @dev An interface for a dividend-paying token contract.
interface DividendPayingTokenInterface {
  /// @notice View the amount of dividend in wei that an address can withdraw.
  /// @param _owner The address of a token holder.
  /// @return The amount of dividend in wei that `_owner` can withdraw.
  function dividendOf(address _owner) external view returns(uint256);

  /// @notice Distributes ether to token holders as dividends.
  /// @dev SHOULD distribute the paid ether to token holders as dividends.
  ///  SHOULD NOT directly transfer ether to token holders in this function.
  ///  MUST emit a `DividendsDistributed` event when the amount of distributed ether is greater than 0.
  function distributeDividends() external payable;

  /// @notice Withdraws the ether distributed to the sender.
  /// @dev SHOULD transfer `dividendOf(msg.sender)` wei to `msg.sender`, and `dividendOf(msg.sender)` SHOULD be 0 after the transfer.
  ///  MUST emit a `DividendWithdrawn` event if the amount of ether transferred is greater than 0.
  function withdrawDividend() external;

  /// @dev This event MUST emit when ether is distributed to token holders.
  /// @param from The address which sends ether to this contract.
  /// @param weiAmount The amount of distributed ether in wei.
  event DividendsDistributed(
    address indexed from,
    uint256 weiAmount
  );

  /// @dev This event MUST emit when an address withdraws their dividend.
  /// @param to The address which withdraws ether from this contract.
  /// @param weiAmount The amount of withdrawn ether in wei.
  event DividendWithdrawn(
    address indexed to,
    uint256 weiAmount
  );
}
```

OPTIONAL functions for a dividend-paying token contract:

```solidity
pragma solidity ^0.5.0;


/// @title Dividend-Paying Token Optional Interface
/// @author Roger Wu (https://github.com/roger-wu)
/// @dev OPTIONAL functions for a dividend-paying token contract.
interface DividendPayingTokenOptionalInterface {
  /// @notice View the amount of dividend in wei that an address can withdraw.
  /// @param _owner The address of a token holder.
  /// @return The amount of dividend in wei that `_owner` can withdraw.
  function withdrawableDividendOf(address _owner) external view returns(uint256);

  /// @notice View the amount of dividend in wei that an address has withdrawn.
  /// @param _owner The address of a token holder.
  /// @return The amount of dividend in wei that `_owner` has withdrawn.
  function withdrawnDividendOf(address _owner) external view returns(uint256);

  /// @notice View the amount of dividend in wei that an address has earned in total.
  /// @dev accumulativeDividendOf(_owner) = withdrawableDividendOf(_owner) + withdrawnDividendOf(_owner)
  /// @param _owner The address of a token holder.
  /// @return The amount of dividend in wei that `_owner` has earned in total.
  function accumulativeDividendOf(address _owner) external view returns(uint256);
}
```

## Implementation

Our implementation of a mintable dividend-paying ERC20 token: https://github.com/Roger-Wu/DividendPayingToken

## Backwards Compatibility

The interface is designed to be compatible with ERC20 (#20) and ERC1400 (#1411) security token standard.

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

**Comment by jtremback** [comment]
I'm interested in this- is this the only/best dividend token design?

**Comment by jtremback** [comment]
I'm having a hard time understanding what the difference is between

`function dividendOf(address _owner) external view returns(uint256);`

and

`function withdrawableDividendOf(address _owner) external view returns(uint256);`

especially since one just calls the other in your implementation

**Comment by Roger-Wu** [comment]
> I'm interested in this- is this the only/best dividend token design?

Hi, @jtremback 
Good to know that you are interested!

I think you are referring to our implementation of a dividend-paying token rather than the interface design of it.
It's of course not the only implementation of such tokens, but we think it's a good enough one in most use cases.
It costs the same amount of gas to distribute and withdraw dividends no matter how many token holders there are.
And the error between the actual value and the ideal value of a dividend is only several weis.

We are not the one who invented this dividend distribution mechanism.
We first saw it in [the PoWH3D token contract](https://etherscan.io/address/0xB3775fB83F7D12A36E0475aBdD1FCA35c091efBe#code).
We re-wrote it to make it cleaner and more reusable.

**Comment by Roger-Wu** [comment]
> I'm having a hard time understanding what the difference is between
> 
> `function dividendOf(address _owner) external view returns(uint256);`
> 
> and
> 
> `function withdrawableDividendOf(address _owner) external view returns(uint256);`
> 
> especially since one just calls the other in your implementation

Yes, they actually do the same thing.

We think that `dividendOf` may be ambiguous.
Some may think that it means "all the dividends you've got", including the withdrawn part and the not yet withdrawn part, while some may think it is only the withdrawable part.
That's why we added the three OPTIONAL viewing functions.

We definitely need more suggestions from the community about naming things.
Please leave a comment if you have any!

**Comment by jo-tud** [comment]
Hi Roger, is there a special reason why you chose 2**64 as magnitude in the implementation?

**Comment by jo-tud** [comment]
> We first saw it in [the PoWH3D token contract](https://etherscan.io/address/0xB3775fB83F7D12A36E0475aBdD1FCA35c091efBe#code).

Here's some additional foundational work for this appraoch: https://medium.com/@weka/dividend-bearing-tokens-on-ethereum-42d01c710657 by @Arachnid

**Comment by Roger-Wu** [comment]
> Hi Roger, is there a special reason why you chose 2**64 as magnitude in the implementation?

### Short Answer

We simply followed what PoWH3D did.
Something between 2\**128 and 2\**165 may be a better choice for `magnitude`.

### Long Answer

In this implementation, the constraints of `magnitude` is made by these lines:

```js
magnifiedDividendPerShare = magnifiedDividendPerShare.add(
  (msg.value).mul(magnitude) / totalSupply()
);

```

```js
function accumulativeDividendOf(address _owner) public view returns(uint256) {
  return magnifiedDividendPerShare.mul(balanceOf(_owner)).toInt256Safe()
    .add(magnifiedDividendCorrections[_owner]).toUint256Safe() / magnitude;
}
```

We want
1. `(msg.value).mul(magnitude)` never overflows.
1. `magnifiedDividendPerShare` never overflows.
1. `magnifiedDividendPerShare.mul(balanceOf(_owner)).toInt256Safe()` never overflows.
1. the remainder of `(msg.value).mul(magnitude) / totalSupply()` is as small as possible compared to `magnitude`.

To sum up, we want to maximize `magnitude` while `(total received ether in wei) * magnitude < 2**255`.

Say, if we know that we will never receive > 1 billion ETH, or about 2\**90 wei, then 2\**165 is a good choice for `magnitude`.

**Comment by Roger-Wu** [comment]
renamed `payAndDistributeDividends` to `distributeDividends`

**Comment by allmysmarts** [comment]
Hi Roger,   Whenever holder's balance is changed(by transfer, mint or burn), his dividend will be kept same as before?
(by the corrections table)

**Comment by mosesanarfi21** [comment]
hi roger , im trying to use this contract and i dont understand it can you help me

**Comment by Roger-Wu** [comment]
> @allmysmarts 
> Whenever holder's balance is changed(by transfer, mint or burn), his dividend will be kept same as before? (by the corrections table)

Yes.

> @mosesanarfi21
> im trying to use this contract and i dont understand it can you help me

Sorry, I'm no longer maintaining the contracts.
You need to figure it out on your own.

**Comment by Zachinquarantine** [comment]
Did this ever get turned into an official Draft EIP or no?

**Comment by Roger-Wu** [comment]
No.

**Comment by chychkoi** [comment]
@Roger-Wu currently this only ETH, is there any possibility this can be used with erc20 token?

**Comment by jo-es** [comment]
we made [this](https://github.com/atpar/funds-distribution-token/blob/master/contracts/extensions/FDT_ERC20Extension.sol) a while ago which is closely aligned with EIP-1726. which is part of a reference implementation of [EIP-2222](https://github.com/ethereum/EIPs/issues/2222). though keep in mind that all of this code is unaudited and outdated.

**Comment by ares9515** [comment]
Hello Roger, if manually send ether to the contract, withdrawableDividendOf will increase?

**Comment by MicahZoltu** [comment]
If continued discussion about this idea is desired, it should be moved to Ethereum Magicians forum or a formal EIP draft should be created.

**Comment by infinity-global** [comment]
essa soluçao tambem pode ser usada em blockhain da binance

**Comment by italoHonoratoSA** [comment]
> essa soluçao tambem pode ser usada em blockhain da binance




The BSC blockchain is based on Ethereum's EVM. So yes, the code syntax and logic works.

**Comment by Alexander-Herranz** [comment]
I have published a modification (updated to Solidity 0.8.1) to be able to share ERC20 tokens as dividends, instead of native tokens:
https://github.com/Alexander-Herranz/ERC20DividendPayingToken

**Comment by heddhshan** [comment]
https://github.com/heddhshan/DividendTokenAndDex/tree/main/Contract， 
I have coded a dividend token and a decentralized exchange for dividend tokens,  some change , i hope IDividendToken is EIP.

**Comment by cyber-storm-200712** [comment]
## EverGrown Coin (Typical Sample of Dividend Paying Token)
https://bscscan.com/address/0xC001BBe2B87079294C63EcE98BdD0a88D761434e#code

**Comment by Lucienest** [comment]
> [https://github.com/heddhshan/DividendTokenAndDex/tree/main/Contract，](https://github.com/heddhshan/DividendTokenAndDex/tree/main/Contract%EF%BC%8C) I have coded a dividend token and a decentralized exchange for dividend tokens, some change , i hope IDividendToken is EIP.

if the comments were in English that'd be helpful @heddhshan

**Comment by internetmoneydev** [comment]
does anyone have an analysis on this pattern vs something like https://eips.ethereum.org/EIPS/eip-4626 ?

**Comment by GuyPhrikshin** [comment]
The contract function "swapTokensForExactTokens" reverted with the following reason: Joe: K Copy the error

**Comment by saurabh-lodha-16** [comment]
@Roger-Wu 
Amazing work on this standard and for your contributions to this.
I had a task exactly similar to this i.e. creating a Dividend Tracker contract and I stumbled upon this.

I followed a lot of threads around this and was quite confused on what finally happened with this standard and if its still safe to use in production.

There is no other way for me to reach out to you so thought to add this here. Would love to know your suggestions on going around this. Thanks
